plugins/legion-go2/fu-legion-go2-device.c

```diff
@@ -0,0 +1,671 @@
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+#include "config.h"
+
+#include "fu-legion-go2-device.h"
+#include "fu-legion-go2-struct.h"
+
+#define FU_LEGION_GO2_DEVICE_VID       0x17EF
+#define FU_LEGION_GO2_DEVICE_PID_BEGIN 0x61EB
+#define FU_LEGION_GO2_DEVICE_PID_END   0x61EE
+
+#define FU_LEGION_GO2_DEVICE_IO_TIMEOUT            500
+#define FU_LEGION_GO2_DEVICE_BUSY_WAIT_TIMEOUT     (60*1000)
+#define FU_LEGION_GO2_DEVICE_READ_RESPONSE_TIMEOUT 2000
+
+#define FU_LEGION_GO2_DEVICE_FW_SIGNED_LENGTH 384
+#define FU_LEGION_GO2_DEVICE_FW_ID_LENGTH     4
+#define FU_LEGION_GO2_DEVICE_FW_PACKET_LENGTH 32
+#define FU_LEGION_GO2_DEVICE_FW_REPORT_LENGTH 64
+
+
+struct _FuLegionGo2Device {
+	FuHidrawDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuLegionGo2Device, fu_legion_go2_device, FU_TYPE_HIDRAW_DEVICE)
+
+static gboolean
+fu_legion_go2_device_read_response(FuLegionGo2Device *self, guint8 main_id, guint8 sub_id, GByteArray* res)
+{
+	gint64 timeNow = g_get_monotonic_time();
+	while (TRUE)
+	{
+		gint64 timeNext = g_get_monotonic_time();
+		gint64 interval = (timeNext - timeNow) / 1000;
+		if (interval > FU_LEGION_GO2_DEVICE_READ_RESPONSE_TIMEOUT)
+		{
+			g_message("read response timeout: ");
+			break;
+		}
+		if (!fu_udev_device_read(FU_UDEV_DEVICE(self),
+						res->data,
+						res->len,
+						NULL,
+						FU_LEGION_GO2_DEVICE_IO_TIMEOUT,
+						FU_IO_CHANNEL_FLAG_NONE,
+						NULL)) {
+			g_message("execute fu_udev_device_read failed: ");
+			continue;
+		}
+		if (res->data[2] == main_id &&
+		    res->data[3] == sub_id)
+		{
+			return TRUE;
+		}
+	}
+
+	return FALSE;
+}
+
+static gboolean
+fu_legion_go2_device_read_response_ex(FuLegionGo2Device *self, guint8 main_id, guint8 sub_id, guint8 id, GByteArray* res)
+{
+	gint64 timeNow = g_get_monotonic_time();
+	while (TRUE)
+	{
+		gint64 timeNext = g_get_monotonic_time();
+		gint64 interval = (timeNext - timeNow) / 1000;
+		if (interval > FU_LEGION_GO2_DEVICE_READ_RESPONSE_TIMEOUT)
+		{
+			g_message("read response timeout: ");
+			break;
+		}
+		if (!fu_udev_device_read(FU_UDEV_DEVICE(self),
+						res->data,
+						res->len,
+						NULL,
+						FU_LEGION_GO2_DEVICE_IO_TIMEOUT,
+						FU_IO_CHANNEL_FLAG_NONE,
+						NULL)) {
+			g_message("execute fu_udev_device_read failed: ");
+			continue;
+		}
+		if (res->data[2] == main_id &&
+		    res->data[3] == sub_id &&
+		    res->data[4] ==  id)
+		{
+			return TRUE;
+		}
+	}
+
+	return FALSE;
+}
+
+static gboolean
+fu_legion_go2_device_upgrade_start(FuLegionGo2Device *self, GByteArray *req, GByteArray* res, GError **error)
+{
+	if (req != NULL) {
+		if (!fu_udev_device_write(FU_UDEV_DEVICE(self),
+					  req->data,
+					  req->len,
+					  FU_LEGION_GO2_DEVICE_IO_TIMEOUT,
+					  FU_IO_CHANNEL_FLAG_NONE,
+					  error)) {
+			g_message("send start command failed: ");
+			return FALSE;
+		}
+	}
+	if (res != NULL) {
+		if (!fu_legion_go2_device_read_response(self, 0x53, 0x11, res))
+		{
+			g_message("read start response failed: ");
```

Comment:
you need to pass error to fu_legion_go2_device_read_response, and `g_prefix_error_literal(error, "read start response failed: ");` and _then_ return FALSE.